app/graphql/types/rule.rb

```diff
@@ -14,37 +14,66 @@ class Rule < Types::BaseObject
     field :severity, String, null: false
     field :remediation_available, Boolean, null: false
     field :profiles, [::Types::Profile], null: true
-    field :identifier, ::Types::RuleIdentifier, null: true
-    field :references, [::Types::RuleReference], null: true,
-                                                 extras: [:lookahead]
+    field :identifier, String, null: true
+    field :references, String, null: true
     field :compliant, Boolean, null: false do
-      argument :system_id, String, 'Is a system compliant?', required: true
+      argument :system_id, String, 'Is a system compliant?',
+               required: false
+      argument :profile_id, String, 'Is a system compliant with this profile?',
+               required: false
     end
 
     def compliant(args = {})
-      return false unless system_id(args) && profile_id(args)
+      system_id(args) &&
+        profile_id(args) &&
+        %w[pass notapplicable notselected].include?(
+          context[:rule_results][object.id][profile_id(args)]
+        )
+    end
 
-      object.compliant?(
-        ::Host.find(system_id(args)),
-        ::Profile.find(profile_id(args))
-      )
+    def references
+      if context[:"rule_references_#{object.id}"].nil?
+        ::CollectionLoader.for(::Rule, :rule_references)
+                          .load(object).then do |references|
+          references.map { |ref| [ref.href, ref.label] }.to_json
+        end
+      else
+        references_from_context
+      end
     end
 
-    def references(lookahead:)
-      selected_columns = lookahead.selections.map(&:name) &
-                         ::RuleReference.column_names.map(&:to_sym)
-      object.references.distinct.reorder('')
-            .pluck(selected_columns).map do |reference|
-        selected_columns.zip([reference].flatten).to_h
+    def references_from_context
+      ::RecordLoader.for(::RuleReference)
+                    .load_many(context[:"rule_references_#{object.id}"])
```

Comment:
Here is the key - if we already have all references loaded here, we can simply load them in one SELECT. 
Before this, if you just have the field defiend as `field :rule_references, [RuleReference]`, GraphQL would do this:
1. Find the profile "foo"
2. Find rule A in profile "foo". 
3. Find all rule references related to rule A, one by one to render them as RuleReference (SELECT * FROM rule_references WHERE id = single rule reference ID)
4. Repeat step 3 until all references are found and rendered.
5. Repeat step 2 until all rules are found

Whereas right now, we already preload nearly everything:

1. Find profile "foo"
2. Batch find all rules in profile "foo" (one SELECT for all rules, even if we query for multiple profiles are queried)
3. Batch find all references in rules found in 2. (one SELECT for all rule references)